### PR TITLE
Extract conversation data into a class, make maps real dicts

### PR DESCRIPTION
**Pull request by crdev**

### CODE REVIEW

File: conversation_data.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
+from node_util import visit_node
+from typing import Dict, List
+
+import proto.conversation_pb2 as conversation_proto
+
+
+def create_node_by_name(
```
Comment (skd) [inline conversation_data.py:7]: If these methods are not meant to be called outside of this module, add a prefix _ to their names.

https://docs.python.org/3.10/tutorial/classes.html#private-variables

File: conversation_data.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+    for node in conversation.node:
+        visit_node(node, updater)
+    return node_by_name
+
+
+def create_keyboard_options(node_by_name) -> Dict[str, List[str]]:
+    keyboard_by_name = dict()
+    for name in node_by_name:
+        if len(node_by_name[name].link) > 0:
```
Comment (skd) [inline conversation_data.py:23]: Nit: invert this:

```
if len(node_by_name[name].link) == 0:
  continue
```

File: conversation_data.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+                elif len(link.branch.name) > 0:
+                    options.append([link.branch.name])
+            keyboard_by_name[name] = options
+    return keyboard_by_name
+
+
+class ConversationData:
+
+    def __init__(self, conversation: conversation_proto.Conversation):
+        self._node_by_name = create_node_by_name(conversation)
```
Comment (skd) [inline conversation_data.py:37]: Nit: I think you can specify types here too:

`self._node_by_name: Dict[str, conversation_proto.ConversationNode] = ...`

File: conversation_data.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+    def updater(node):
+        node_by_name[node.name] = node
+
+    for node in conversation.node:
+        visit_node(node, updater)
+    return node_by_name
+
+
+def create_keyboard_options(node_by_name) -> Dict[str, List[str]]:
```
Comment (skd) [inline conversation_data.py:20]: Isn't the return type List[List[str]]?

File: conversation_data.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+class ConversationData:
+
+    def __init__(self, conversation: conversation_proto.Conversation):
+        self._node_by_name = create_node_by_name(conversation)
+        self._keyboard_by_name = create_keyboard_options(self._node_by_name)
+
+    def node_by_name(self, name: str) -> conversation_proto.ConversationNode:
+        return self._node_by_name.get(name)
+
```
Comment (skd) [inline conversation_data.py:42]: Isn't the return type List[List[str]]?

**Review by skd — APPROVED** [review]

File: conversation_data.py (hunk 4 above)
Comment (crdev) [reply conversation_data.py:20]: Thanks, overlooked

File: conversation_data.py (hunk 5 above)
Comment (crdev) [reply conversation_data.py:42]: Nope :) That's keyboard_by_name().

**Pull request merged**